### ISSUE TITLE
[$1000] Implement new Google Places API

### ISSUE THREAD

**Issue by rafecolton** [Expensify/App#46771]
Coming from https://github.com/Expensify/App/issues/45432, please implement the new Places API as described in [this comment](https://github.com/Expensify/App/issues/45432#issuecomment-2263739363).

We'll follow the usual process of starting with a detailed proposal, but I will assign @rojiphil directly as the Contributor here since he has the most context.

I will be the CME and will also help with getting the API key updated/installed.

----

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~015599eafb1cb0d217</li>
        <li>Upwork Job ID: 1819786402974465675</li>
        <li>Last Price Increase: 2024-08-03</li>
<li>Automatic offers: </li>
<ul>
<li>mkhutornyi | Reviewer | 103384006</li>
<li>rojiphil | Contributor | 103384007</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @Christinadobrzyn</details>

**Comment by rafecolton** [comment]
Not sure if the upwork job was created, going to try re-adding the label

**Comment by rafecolton** [comment]
Per discussion in the linked issue, I'm assigning @rojiphil as the Contributor. @rojiphil, please provide a detailed proposal when you are able for @mkhutornyi to review. Thanks!

**Comment by rojiphil** [comment]
Thanks @rafecolton for assigning me here. I am looking into this and expect to revert in next couple of days.

**Comment by rojiphil** [comment]
> Any update on when you will be delivering the first proposal @rojiphil?

Almost there. I will share the update today.

**Comment by rojiphil** [comment]
@rafecolton @mkhutornyi  A draft PR with the proposed implementation details is available in PR description [here](https://github.com/Expensify/App/pull/47085). Please have a look. We can take this further based on your feedback.

**Comment by rafecolton** [comment]
Thanks for the initial writeup. @mkhutornyi please take a look at the proposal and let me know your thoughts.

In order to keep the discussion all in one place, @rojiphil will you please transfer the proposal here? (Feel free to edit your comment above to include the proposal details.)

I have a few initial questions and comments:
1. Are you able to test the current API by making a real API request through our web proxy? If not, how do you plan to test? (i.e. Will you need support from an internal engineer?)
2. Can you please add some detail around what specific changes will need to be made to the BE? 
    * The current implementation in the web API essentially proxies the request to `https://maps.googleapis.com/maps/api`, passing along the query parameters from the front-end after injecting/replacing the api key. 
    * Please detail any changes that will be made to the URL, http method, etc. Please ensure this design will remain backwards-compatible. (We either need to add a new proxy endpoint or update the existing one to differentiate between the requests and support both types.)
    * Please also detail what additional scopes need to be added to the API key, or if a new key is required. If it's possible to add scopes to the existing key rather than generate a new one, I think that would be preferred, as it will simplify the BE code a bit.
3. I will be OOO for the rest of the month starting tomorrow, so @Christinadobrzyn can source another engineer to help you test if you are not able to test yourself. That engineer may have to recruit somebody to generate and install the new key or add scopes to the existing one.

**Comment by rojiphil** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

This PR implements the [scope](https://github.com/Expensify/App/issues/45432#issuecomment-2263794353) limited to the places migration as used in NewDot.

### What is the root cause of that problem?

There is no root cause as this is an improvement.

### What changes do you think we should make in order to solve the problem?

Dependencies:
1. Changes in `react-native-google-places-autocomplete` library which is implemented in [this PR](https://github.com/FaridSafi/react-native-google-places-autocomplete/pull/943)
2. API Key updation for `Places API (New)` in Google Cloud 
3. BE changes to support API changes for `Autocomplete (New)` and `Place Details(New)`

The implementation details for the relevant steps as mentioned [here](https://github.com/Expensify/App/issues/45432#issuecomment-2263739363) are as follows:

Let us introduce a `isNewPlacesAPI` property in `GooglePlacesAutocomplete` component [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/e3676609bae828a98ce4a1de44bf6312a968a4e5/GooglePlacesAutocomplete.js#L1006C3-L1006C17) to introduce support for `Places (New)` api. By default, we can set this to `false` [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/e3676609bae828a98ce4a1de44bf6312a968a4e5/GooglePlacesAutocomplete.js#L1006C3-L1006C17) so that the current implementation of `Places` api is used by default. This way, we do not break existing systems using the old `Places` API. Further, to make use of the `Places API (New)`, the FE can send `isNewPlacesAPI` [here](https://github.com/rojiphil/App/blob/6425853eadcd8a6c19a183217d97c102219b7416/src/components/AddressSearch/index.tsx#L476). 

1. Use HTTP POST instead of HTTP GET for Autocomplete (New) API uses as mentioned [here](https://developers.google.com/maps/documentation/places/web-service/place-autocomplete#place_autocomplete_requests).

For `Autocomplete (New)` request, we can send a HTTP POST request [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/e3676609bae828a98ce4a1de44bf6312a968a4e5/GooglePlacesAutocomplete.js#L609) and also set the new url for POST i.e. `${url}/v1/places:autocomplete`. Further, we can send the request `body` with the supported parameters [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/e3676609bae828a98ce4a1de44bf6312a968a4e5/GooglePlacesAutocomplete.js#L624C1-L630C12).
In addition, we also need to update the API for `Place Details (New)` with url `${url}/v1/places/${rowData.place_id}` [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/e3676609bae828a98ce4a1de44bf6312a968a4e5/GooglePlacesAutocomplete.js#L334)

2. Handle the format changes in response structure for Autocomplete (New) API as mentioned [here](https://developers.google.com/maps/documentation/places/web-service/place-autocomplete#:~:text=5000%20meter%20radius%3A-,%7B,-%22suggestions%22%3A)

Format changes include changing the request parameters to `languageCode`, `includedRegionCodes` and `locationBias` [here](https://github.com/rojiphil/App/blob/6425853eadcd8a6c19a183217d97c102219b7416/src/components/AddressSearch/index.tsx#L99C1-L101C49). For the response though, we need to filter the results and handle the format changes. This is done [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/e3676609bae828a98ce4a1de44bf6312a968a4e5/GooglePlacesAutocomplete.js#L440C1-L461C5) and [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/e3676609bae828a98ce4a1de44bf6312a968a4e5/GooglePlacesAutocomplete.js#L582C1-L589C12)

3. Use the required parameter fieldmask for Place Details (New) API as mentioned [here](https://developers.google.com/maps/documentation/places/web-service/place-details#fieldmask). This way we can control the billing by including what we need in field mask

Let us introduce a `fields` property in `GooglePlacesAutocomplete` component [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/e3676609bae828a98ce4a1de44bf6312a968a4e5/GooglePlacesAutocomplete.js#L1007) to specify the desired response fields. By default, we can set this to `*` [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/e3676609bae828a98ce4a1de44bf6312a968a4e5/GooglePlacesAutocomplete.js#L1053) so that all the fields are sent back. In our case, we set `fields` to make use of [Location Only (Place Details) SKU](https://developers.google.com/maps/documentation/places/web-service/usage-and-billing#location-placedetails)  thereby reducing our API cost. The specific value of `fields` is set [here](https://github.com/rojiphil/App/blob/6425853eadcd8a6c19a183217d97c102219b7416/src/CONST.ts#L1193).

4. Use session tokens to group autocomplete and place details which will help in reduced billing. This is referenced [here](https://developers.google.com/maps/documentation/places/web-service/place-session-tokens).

To use sessions, we can create a `sessionToken` state [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/e3676609bae828a98ce4a1de44bf6312a968a4e5/GooglePlacesAutocomplete.js#L163) and use `uuid` library to generate v4 uuids for the session. This session token will be used in every `Autocomplete (New)` API [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/e3676609bae828a98ce4a1de44bf6312a968a4e5/GooglePlacesAutocomplete.js#L627). And the session token will be reset to a new uuid [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/e3676609bae828a98ce4a1de44bf6312a968a4e5/GooglePlacesAutocomplete.js#L340) after using it for the last time in the `Place Details (New)` API [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/e3676609bae828a98ce4a1de44bf6312a968a4e5/GooglePlacesAutocomplete.js#L336) 

5. Handle the format changes related to Place object for in response structure of Place Details (New) API as referenced [here](https://developers.google.com/maps/documentation/places/web-service/migrate-response)

Since the `response` has changed `Place Details (New)` API, there will not be any `status`. Hence, we can depend on `id` [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/e3676609bae828a98ce4a1de44bf6312a968a4e5/GooglePlacesAutocomplete.js#L288) and pass the `details` to the `onPress` handler. The FE [saveLocationDetails](https://github.com/rojiphil/App/blob/6425853eadcd8a6c19a183217d97c102219b7416/src/components/AddressSearch/index.tsx#L106) is modified to support the format changes in the `Place Details (New)` API response.

### What alternative solutions did you explore? (Optional)

**Comment by rojiphil** [comment]
> * Are you able to test the current API by making a real API request through our web proxy? If not, how do you plan to test? (i.e. Will you need support from an internal engineer?)

We will need the support of an internal engineer to make the BE changes. 
With the current changes, the BE returns `666` code response. So, to test until we are BE ready, I have injected some test data in the code which will be removed once the BE is ready. As soon as the BE is ready, the FE should work seamlessly though.

> * Can you please add some detail around what specific changes will need to be made to the BE?
>   
>   * The current implementation in the web API essentially proxies the request to `https://maps.googleapis.com/maps/api`, passing along the query parameters from the front-end after injecting/replacing the api key.
>   * Please detail any changes that will be made to the URL, http method, etc. Please ensure this design will remain backwards-compatible. (We either need to add a new proxy endpoint or update the existing one to differentiate between the requests and support both types.)

The specific changes are as follows:

1. Current implementation proxies the request to `https://maps.googleapis.com/maps/api`. The new implementation should proxy the request to `https://places.googleapis.com` for both `Autocomplete API (New)` and `Place Details (New)`. And, as usual, we need to pass along the query parameters from the FE after injecting/replacing the api key.

2. I proposed to add new proxy endpoints. That way, our design would be backward-compatible.

For Autocomplete API (New):
HTTP Method: POST
Query URL begins with /v1/places:autocomplete

For Place Details API (New):
HTTP Method: GET
Query URL begins with /v1/places/${rowData.place_id} where `rowData.place_id` represents the placeId of the place for which details are to be fetched. 

> * Please also detail what additional scopes need to be added to the API key, or if a new key is required. If it's possible to add scopes to the existing key rather than generate a new one, I think that would be preferred, as it will simplify the BE code a bit.

The steps required to enable `Places API (New)` to the API key is mentioned [here](https://developers.google.com/maps/documentation/places/web-service/migrate-overview#enable-new). I think we can update the existing API key itself if it is going to simplify the BE.

**Comment by Christinadobrzyn** [comment]
Hi @mkhutornyi just checking in on this, can you please provide an update on the review? thank you!

**Comment by Christinadobrzyn** [comment]
Awesome thanks for the review @mkhutornyi!

**Comment by thienlnam** [comment]
Looks like Rafe is going to be the CME on this issue, so going to un-assign myself

**Comment by rojiphil** [comment]
> The [proposal](https://github.com/Expensify/App/issues/46771#issuecomment-2276315538) looks great.
> I think we can go ahead with the implemenation.

Awesome. Thanks @mkhutornyi for your review.

@rafecolton When we can expect the API key updation and the handling of APIs in BE? 
Once that is done, I can retest the draft PRs with the changes in BE, remove the test code in library, and move the draft PRs to review so that the maintainer of library can review/merge the updates.

**Comment by rafecolton** [comment]
I'm going OOO for a couple weeks starting tomorrow, so it won't be until I get back.

**Comment by Christinadobrzyn** [comment]
We'll move this to weekly until Rafe is back.

**Comment by thienlnam** [comment]
Looks like it's pending some BE changes, so it will just have to wait till he gets back or another internal engineer picks it up

**Comment by Christinadobrzyn** [comment]
Awesome! Sounds good! thank you for taking a peek!

**Comment by Christinadobrzyn** [comment]
Adding an internal label to this since there are some BE components. We'll pick this back up when Rafe is back!

**Comment by Christinadobrzyn** [comment]
This is on hold for a bit longer while Rafe is away!

**Comment by trjExpensify** [comment]
@rafecolton is there a TL;DR on why we need to do this?

**Comment by rafecolton** [comment]
> is there a TL;DR on why we need to do this?

@trjExpensify **tl;dr** The new new APIs return better data faster for a lower cost. More on what was shared by our Google reps [here](https://github.com/Expensify/App/issues/45432#issue-2409725263)

**Comment by Christinadobrzyn** [comment]
just a heads up - I'm going to be ooo 9/4 - 9/10. I don't think we'll need another BZ teammate on this issue, so I'll just keep myself, but if you do, please reach out in Slack for someone to jump in!

**Comment by rojiphil** [comment]
@rafecolton Gentle bump on taking this forward. Also, can we please move this issue back to `daily`?

**Comment by trjExpensify** [comment]
> > is there a TL;DR on why we need to do this?
> 
> @trjExpensify **tl;dr** The new new APIs return better data faster for a lower cost. More on what was shared by our Google reps [here](https://github.com/Expensify/App/issues/45432#issue-2409725263)

Gotcha, it's pretty light on the upside for Places API there in their message in the respect of something we actually want to use.

>The Places API has also been upgraded and some of those changes can have an impact, you can find about the New Places APIs [here](https://developers.google.com/maps/documentation/places/web-service/op-overview).

Are we saving a lot with this new Places API then?

**Comment by rojiphil** [comment]
> Are we saving a lot with this new Places API then?

I think the savings are significant. Here is a summary of and advantages for making the changes.

Places API (Old) [Reference](https://developers.google.com/maps/documentation/places/web-service/usage-and-billing#query-ac-per-request)
1000 Autocomplete Requests: $2.83
1000 Place Details Requests: $17.00

Places API (New)  [Reference](https://developers.google.com/maps/documentation/places/web-service/session-pricing#pricing-examples)
1000 Autocomplete Requests: $2.83
1000 Place Details Requests: $5.00

Advantages:
1. Flat $12 (17-5) savings for every 1000 Place Details Request as we use `Location` Field Mask in Places API (New). The existing Place Details Request sends all details but we do not need all those details. With the introduction of field mask in Places API (New), we can fetch only what we want and, hence, the savings.
2. Additionally, we are also introducing `session` based implementation which clubs `AutoComplete` requests made in a session as a single `AutoComplete` request from the pricing perspective. So,
1000 `autocomplete` requests without session —>  $2.83
1000 `autocomplete` requests made in 200 sessions (I.e. 5 autocomplete requests per session) will be considered as 200 `autocomplete` requests —> 5x (i.e. 5000 autocomplete requests) such sessions will incur charge of $2.83

**Comment by rafecolton** [comment]
Not huge in terms of prioritization @trjExpensify but in terms of $$$ it will pay for itself 😄

**Comment by rafecolton** [comment]
> @rafecolton Gentle bump on taking this forward. Also, can we please move this issue back to daily?

Thanks for your patience, I'm catching up from OOO this week, should be able to do that Monday

**Comment by rafecolton** [comment]
Thanks for your patience. Getting back to the BE component of this now.

**Comment by rafecolton** [comment]
I enabled the new Places API in GCP, working on supporting it through our web API now

**Comment by rafecolton** [comment]
@rojiphil did some testing locally. Left a small comment on https://github.com/FaridSafi/react-native-google-places-autocomplete/pull/943 and some changes are needed on the App PR. With those small changes, I got the BE working! So I'm putting up that PR now.

Please proceed with getting https://github.com/FaridSafi/react-native-google-places-autocomplete/pull/943 ready for review and https://github.com/Expensify/App/pull/47085 after that 😊

@rojiphil @mkhutornyi Are you able to test on Expensify Staging? If not, are you able to get your own API key so you can test e2e locally?

**Comment by rafecolton** [comment]
Oh I forgot to mention - I made the existing proxy command backwards-compatible, so we won't need a new command/url

**Comment by rafecolton** [comment]
The BE changes were merged and deployed to staging today, so they should be in prod on Monday. We are discussing on the App PR whether or not it will be possible for C/C+ to test on staging.

**Comment by Christinadobrzyn** [comment]
yes, not overdue Melvin.

**Comment by rafecolton** [comment]
Saw you pushed some changes @rojiphil and commented on the PRs. I've been swamped today, so I'll try to get to this tomorrow. Making myself the issue owner now since this is waiting on my review.

**Comment by Christinadobrzyn** [comment]
Just a heads up that I'm going to be travelling for the rest of the week - back on Monday. I'm not going to assign a new BZ teammate since this is being handled at the PR stage.

**Comment by rafecolton** [comment]
No problem, thanks for the heads up. @rojiphil thanks for your patience, I've been swamped the last couple days. Hope to get back to this ASAP.

**Comment by rafecolton** [comment]
Looking at this now

**Comment by rafecolton** [comment]
Marking `Reviewing` as all related PRs are under review currently

**Comment by Christinadobrzyn** [comment]
With `Reviewing`, should we move this from `daily` to `weekly`?

**Comment by rafecolton** [comment]
Let's keep it at `Daily`. In fact, I'll take it out of reviewing and make @rojiphil the issue owner until the final App PR is ready to review.

**Comment by rafecolton** [comment]
Final PR is now up for review! Cool to see that Melvin added the labels automagically ✨

**Comment by rafecolton** [comment]
PR is failing on android without content-type header, but with the header, it produces a CORS error. I dug into that before when I was testing and it was a rabbit hole, but since it seems to be required here, I'll have to dig more. Setting myself as the issue owner now since this is blocked on me. Might not be able to get back to this until next week, have some higher priority issues assigned currently

**Comment by rojiphil** [comment]
@rafecolton Oh! Melvin closed this issue. Can you please reopen this? Or is this not a priority anymore?

**Comment by rafecolton** [comment]
This has never been a super high priority per se, but I would love to get it implemented because it will save us some money and pay for itself. I'm currently looking for a volunteer to take over the backend investigation and get you unblocked there.

**Comment by rojiphil** [comment]
@rafecolton The problem statement looks fine to me but what baffles me is that the FE request with the `Content-type` header succeeds for native(Android/iOS) and Web platforms (i.e. via Chrome/Safari browser) but fails for the Desktop platform with CORs violation as mentioned [here](https://github.com/Expensify/App/pull/47085#issuecomment-2398570019). Does the FE request go through different paths in BE based on the type of FE platform (i.e. Web/Native/Desktop)?

**Comment by rojiphil** [comment]
Ok. We are almost there to get this PR done except for this nasty issue. Hoping that we break the deadlock soon.

**Comment by rafecolton** [comment]
Ideally I'd like to hand this off to a backend external contributor, but they are still ramping up. @flodnv any thoughts?

I'm also curious if we could add a switch to enable the new feature everywhere except android. That would enable us to close this issue out and get you paid, then we could come back and update android when it's a little higher priority. @tgolen what do you think about that?

**Comment by flodnv** [comment]
> Ideally I'd like to hand this off to a backend external contributor, but they are still ramping up. @flodnv any thoughts?

Same as [here](https://expensify.slack.com/archives/C03TQ48KC/p1738096524751869?thread_ts=1737404447.539409&cid=C03TQ48KC) 😄

**Comment by rojiphil** [comment]
> can you please accept the Upwork job here - https://www.upwork.com/nx/wm/offer/108892421

@Christinadobrzyn Offer accepted. Thanks.
